src/headless-graphql-client.ts

```diff
@@ -1,67 +1,68 @@
 import { Address } from "@planetarium/account";
-import {
-    BencodexDictionary,
-    Dictionary,
-    Value,
-    decode,
-    encode,
-} from "@planetarium/bencodex";
+import { BencodexDictionary, Dictionary, decode } from "@planetarium/bencodex";
 import { Currency, FungibleAssetValue } from "@planetarium/tx";
-import axios, { AxiosResponse } from "axios";
+import { Client, cacheExchange, fetchExchange } from "@urql/core";
+import { retryExchange } from "@urql/exchange-retry";
+import {
+    GetAssetTransferredDocument,
+    GetBlockHashDocument,
+    GetBlockIndexDocument,
+    GetGarageUnloadsDocument,
+    GetGenesisHashDocument,
+    GetNextTxNonceDocument,
+    GetTipIndexDocument,
+    GetTransactionResultDocument,
+    StageTransactionDocument,
+} from "./generated/graphql";
 import { IHeadlessGraphQLClient } from "./interfaces/headless-graphql-client";
 import { AssetTransferredEvent } from "./types/asset-transferred-event";
 import { BlockHash } from "./types/block-hash";
 import { GarageUnloadEvent } from "./types/garage-unload-event";
 import { TransactionResult } from "./types/transaction-result";
 import { TxId } from "./types/txid";
 
-function delay(ms: number): Promise<void> {
-    return new Promise((resolve) => {
-        setTimeout(() => {
-            resolve();
-        }, ms);
-    });
-}
-
-interface GraphQLRequestBody {
-    operationName: string | null;
-    query: string;
-    variables: Record<string, unknown>;
-}
-
 export class HeadlessGraphQLClient implements IHeadlessGraphQLClient {
-    private readonly _apiEndpoint: string;
-    private readonly _maxRetry: number;
+    private readonly _client: Client;
+    private readonly _fallbackUrl: URL;
 
     constructor(apiEndpoint: string, maxRetry: number) {
-        this._apiEndpoint = apiEndpoint;
-        this._maxRetry = maxRetry;
+        this._client = new Client({
+            url: apiEndpoint,
+            exchanges: [
+                cacheExchange,
+                retryExchange({
+                    initialDelayMs: 1000,
+                    maxDelayMs: 15000,
+                    randomDelay: true,
+                    maxNumberAttempts: maxRetry,
+                    retryWith: (error, operation) => {
+                        console.error(error.message);
+                        // https://formidable.com/open-source/urql/docs/basics/errors/
+                        // This automatically distinguish, log, process Network / GQL error.
+                        if (error.networkError) {
+                            const context = {
+                                ...operation.context,
+                                url: this._fallbackUrl.toString(),
```

Comment:
It seems never initialized. Is it okay to call `undefined.toString()`?

Reply:
Temporarily Fixed until fallback url provider implemented.